qutip/entropy.py

```diff
@@ -244,24 +243,42 @@ def _entropy_relative(rho, sigma, base=e, sparse=False):
     rel_ent : float
         Value of relative entropy.
 
+    References
+    ----------
+
+    See Nielsen & Chuang, "Quantum Computation and Quantum Information",
+    Section 11.3.1, pg. 511 for a detailed explanation of quantum relative
+    entropy.
     """
-    if rho.type != 'oper' or sigma.type != 'oper':
-        raise TypeError("Inputs must be density matrices..")
-    # sigma terms
-    svals = sp_eigs(sigma.data, sigma.isherm, vecs=False, sparse=sparse)
-    snzvals = svals[svals != 0]
+    if not rho.isoper or not sigma.isoper:
+        raise TypeError("Inputs must be density matrices.")
     if base == 2:
-        slogvals = log2(snzvals)
+        log_base = log2
     elif base == e:
-        slogvals = log(snzvals)
+        log_base = log
     else:
         raise ValueError("Base must be 2 or e.")
-    # rho terms
-    rvals = sp_eigs(rho.data, rho.isherm, vecs=False, sparse=sparse)
-    rnzvals = rvals[rvals != 0]
-    # calculate tr(rho*log sigma)
-    rel_trace = float(real(sum(rnzvals * slogvals)))
-    return -entropy_vn(rho, base, sparse) - rel_trace
+    # S(rho || sigma) = sum_i(p_i log p_i) - sum_ij(p_i P_ij log q_i)
+    #
+    # S is +inf if the kernel of sigma (i.e. svecs[svals == 0]) has non-trivial
+    # intersection with the support of rho (i.e. rvecs[rvals != 0]).
+    rvals, rvecs = sp_eigs(rho.data, rho.isherm, vecs=True, sparse=sparse)
+    svals, svecs = sp_eigs(sigma.data, sigma.isherm, vecs=True, sparse=sparse)
+    nzrvals = rvals[rvals != 0]
+    # Calculate S
+    S = sum(nzrvals * log_base(nzrvals))
+    for i in range(len(rvals)):
+        for j in range(len(svals)):
+            P_ij = (
+                dot(rvecs[i], svecs[j].conjugate()) *
+                dot(svecs[j], rvecs[i].conjugate())
+            )
+            if svals[j] == 0 and (rvals[i] * P_ij != 0):
+                # kernel of sigma intersects support of rho
+                return inf
+            if svals[j] != 0:
+                S -= rvals[i] * P_ij * log_base(svals[j])
+    return real(S)
```

Comment:
The call to `real(S)` looks unnecessary (assuming that it's only ever Hermitian matrices that can be input) - the eigenvalues should be floats not complex already, and `P_ij` should be real itself, though I suppose depending on how you find it, it might not be typed as such.

Reply:
Also, given that we're constructing this by subtracting from a start value, should we take special care to ensure that we never return a value less than 0?  I can imagine it could screw up plenty of people's data processing in loops if we let small negative values through, and with floating-point subtraction that's entirely possible.

Reply:
`S` is a complex number (with zero imaginary part), `real(S)` is a float. Since the relative entropy is always a positive real number, I think `real(S)` makes sense, and agree that clamping to zero makes sense. Will add.

Reply:
I didn't say it very well originally, but my point was that there's no reason `S` _should_ be a complex number at any point in the calculation, right?  The call to `real` made it look like something was wrong to me while I flicked through it.

If the inputs are allowed to be non-Hermitian, then S would in general have a non-zero imaginary component, so the call to `real` doesn't make sense, and if they _are_, then `rvals` and `svals` are both of type `double`, and `P_ij` should be of type `double` in all cases anyway, so S should be too, right?

Reply:
Yes, all the eigenvalues and `P_ij` should be real already. I could convert / raise an error earlier?

Reply:
I think the intention is probably clearer if we do the real-ness tests when the eigenvalues are generated?  That's also approximately how you'd test for a valid PT matrix, I think, too - if the eigenvalues are all within tolerance of a real value then it's valid to try and continue, and we can cast them down to real then.

I don't know how much effort you want to put into non-Hermitian dynamics right now, though - I don't think very much of the library supports it.  If that's the case, then the function could also just fail out if `not rho.isherm or not sigma.isherm`.

Reply:
I've added the checks for non-real eigenvalues and converted the eigenvalues to floats immediately after the checks.